Turn rays, grab, gaze pointers on and off at runtime from code

julenka opened the pull request:
## Overview
We have had many requests for how to disable the far interaction (line pointer, hand rays, etc) at runtime. For example, this [stackoverflow question](https://stackoverflow.com/questions/56248329/how-to-turn-off-motion-controller-line-pointer/56435895#56435895).

The default way to do this is not obvious, and requires creating a custom mediator. Because this is such a commonly requested feature, we should just make it easy to do this when using the default MRTK input system.

This change provides a way for people using the default `FocusProvider` and `DefaultPointerMediator` to dynamically enable and disable pointers at runtime. Pointers can be enabled / disabled per handedness as well, allowing you to, for example, turn off hand rays for the right hand only.

Here is an example of dynamically turning different pointers on / off:

![pointerOnOff](https://user-images.githubusercontent.com/168492/64627539-ae11a500-d3a4-11e9-8513-d8cfed727c5f.gif)


## Changes
- Fixes: #4954
- Introduce `PointerBehavior` to describe how pointer should behave: On, Off, Default (use default MRTK behavior)
- Provide API in `FocusProvider` to set the pointer behavior for a pointer type and handedness
- Provide convenience methods in `PointerUtils` to set common pointer behaviors
- Add tests
- Add example scene
- Add documentation

## Old way to turn pointers on off
See this [stackoverflow question](https://stackoverflow.com/questions/56248329/how-to-turn-off-motion-controller-line-pointer/56435895#56435895), it requires creating custom pointer mediator.

## New way to turn pointers on / off

```csharp
// Turn off all hand rays
PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Off);

// Restore default behavior for rays (on if not near something grabbable)
PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Behavior);

// Turn off hand rays for the right hand only
PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Off, Handedness.Right);

// Turn off the gaze pointer
PointerUtils.SetGazePointerBehavior(PointerBehavior.Off);
```

## Verification
- TODO: Verify on device (HL1, HL2)

julenka commented:
Still TO DO in this PR:
- [ ] Verify works on HL1, HL2
- [ ] Add tests for gaze, grab, poke pointers, and common modes.

Troy-Ferrell commented on Assets/MixedRealityToolkit.Services/InputSystem/PointerBehavior.cs:
```diff
@@ -0,0 +1,33 @@
+    public enum PointerBehavior
+    {
+        /// <summary>
+        /// Pointer is always on, regardless of what other pointers are active.
```
This seems troublesome. I feel like a lot of developers are just going to set On/Off but really most will want the scenario of Off vs Default. Having a pointer On seems dangerous because we have all of that custom logic in PointerMediator to do special steps we think are good

I'm not exactly sure how I would fix this. Maybe at least make Default = 0 since that is the default we want

Troy-Ferrell replied:
Following up on this, I know it's annoying, but it might also be good to change PointerOnOff example scene content to be like PointerBehaviorExample instead. Again the problem is people will generally not want On/off but Default/Off

Troy-Ferrell replied:
Also now seeing the PR and understanding everything, I was an idiot before. These should be AlwaysOn and AlwaysOff enums. Sorry about that

julenka replied:
Updated to the following enum:

AlwaysOn,
AlwaysOff,
Default

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerOnOffTests.cs:
```diff
@@ -0,0 +1,122 @@
+    /// <summary>
+    /// Verify that pointers can be turned on and offvia FocusProvider.SetPointerBehavior
+    /// </summary>
+    public class PointerOnOffTests : BasePlayModeTests
```
nit: Can you rename to PointerBehaviorTests? Seems more accurate

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerOnOffTests.cs:
```diff
@@ -0,0 +1,122 @@
+            public bool? LinePointerEnabled { get; set; }
+            public bool? SpherePointerEnabled { get; set; }
+            public bool? PokePointerEnabled { get; set; }
+            public bool? GazePointerEnabled { get; set; }
```
why use bool? when you have PointerBehavior for the three states?

julenka replied:
because I want to test IsInteractionEnabled of a specific pointer. Otherwise I'm just testing if a setter works. Instead, I'm searching for pointers and actually testing if they are enabled.

Troy-Ferrell commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
```diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
 
 This is handled by the [`DefaultPointerMediator`](xref:Microsoft.MixedReality.Toolkit.Input.DefaultPointerMediator),
-which is responsible for determining which pointers based on the state of all pointers. One of the key things this does is [disable far pointers when a near pointer is close to an object](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs#L127)
+which is responsible for determining which pointers based on the state of all pointers. One of the key things this does is [disable far pointers when a near pointer is close to an object](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs#L127).
```
" which pointers based on" => "which pointers are active based.."

Troy-Ferrell commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
```diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
 It's possible to provide an alternate implementation of the pointer mediator by changing the [`PointerMediator`](xref:Microsoft.MixedReality.Toolkit.Input.MixedRealityPointerProfile.PointerMediator) property on the pointer profile.
 
+### How to turn pointers on and off in MRTK using PointerBehavior
+Because the pointer mediator runs every frame, it ends up controlling the active / inactive state of all pointers. Therefore, if you set a pointer's IsInteractionEnabled property in code, it will get overwritten by the pointer mediator every frame. Instead, you can specify the [`PointerBehavior`](xref:Microsoft.MixedReality.Toolkit.Inputf.PointerBehavior) to control whether pointers should be on  ...
```
Should IsInteractionEnabled be protected/private set then?

Troy-Ferrell replied:
nit: Remove yourself, you etc

julenka replied:
That would be a breaking change. Probably for the better since people really , but could break people. @davidkline-ms @wiwei woudl this be an acceptable change, to remove IsInteractionEnabled from being settable publicly?

wiwei replied:
I think it would need to be internal - as a building block (i.e. as a capability that the pointer provides) having IsInteractionEnabled makes sense to be public so that higher level building blocks (of which other MRTK things are one such example) can set/toggle them. 

The value here seems okay I can think of a number of other cases like this, and I don't think we should just do this in a one-off basis.

wiwei replied:
So I would vote no here.

julenka replied:
Resolving this, I won't be changing it.

Troy-Ferrell commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
````diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
+PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Off, Handedness.Right);
+```
+
+The following code will return hand rays to their default behavior (on unless a hand is near a grabbable) in MRTK:
````
I would remove the (on unless...), distracting comment

Troy-Ferrell commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
````diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
+PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Default);
+```
+
+The following code will force hand rays to be on, regardless of if near a grabbable:
````
regardless of => regardless if

Troy-Ferrell commented on Documentation/Input/Pointers.md:
```diff
@@ -245,6 +245,32 @@ The [PointerResultExample scene](https://github.com/microsoft/MixedRealityToolki
 <img src="../../Documentation/Images/Input/PointerResultExample.png" style="max-width:100%;">
 
+### Turn pointers on and off
+To turn pointers on and off, set the [`PointerBehavior`]() for a given pointer type via [`PointerUtils`](). See [How to turn pointers on and off in MRTK using PointerBehavior]() for more information.
```
Again, emphasize more to control pointer behavior instead of this focus on the duality of on/off

julenka replied:
Renamed to "disable pointers"

Troy-Ferrell commented on Documentation/Input/Pointers.md:
```diff
@@ -245,6 +245,32 @@ The [PointerResultExample scene](https://github.com/microsoft/MixedRealityToolki
+PointerUtils.SetGazePointerBehavior(PointerBehavior.Off);
+
+// Set the behavior to match HoloLens 1
+// Note, if on HoloLens 2, must make the GGV pointer show up for articulated hands. See TurnPointersOnOffExample in the MRTK Examples package
```
I really like that your example shows the typical setup for HL1 vs HL2 vs VR etc. 

Either 1) in this PR we should updated Pointers.md to talk about which pointers are typically valid on which platforms? or 2) file a documentation bug to do something similar later

julenka replied:
Thanks! I added a table

wiwei replied:
Table looks great!

---
In reply to: [323377626](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5920#discussion_r323377626) [](ancestors = 323377626)

wiwei commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TunrPointersOnOff Configuration Profile.asset:
```diff
@@ -0,0 +1,55 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
```
I don't think we usually have spaces in our assets (i.e. there's spaces between Configuration/Profile/TunrPointersOnOff.

Also I think there's a misspelling in TunrPointersOnOff -> TurnPointersOnOff

wiwei replied:
Same general comment on spaces for all of the newly added assets.

wiwei commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TurnPointersOnOff.cs:
```diff
@@ -0,0 +1,134 @@
+    /// </summary>
+    public class TurnPointersOnOff : MonoBehaviour
+    {
+        List<Tuple<Type, Interactable>> pointerToggles = new List<Tuple<Type, Interactable>>();
```
Please add a comment on what this tuple is for (i.e. an example of something that it would store)

julenka replied:
I removed that, in favor of hooking up assets.

wiwei commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TurnPointersOnOff.cs:
```diff
@@ -0,0 +1,134 @@
+
+        private void PopulatePointerToggleList()
+        {
+            foreach (var icl in FindObjectsOfType<Interactable>())
```
```suggestion
            foreach (var interactable in FindObjectsOfType<Interactable>())
```

julenka replied:
Removed the prototype code.

wiwei commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TurnPointersOnOff.cs:
```diff
@@ -0,0 +1,134 @@
+        void Start()
+        {
+            IMixedRealityCapabilityCheck capabilityChecker = CoreServices.InputSystem as IMixedRealityCapabilityCheck;
+            if (capabilityChecker != null)
```
@davidkline-ms just to check, is it the case that the input system should have deterministically started up by any arbitrary monobehavior's Start() function?

I know there were some others that tried accessing things but had to wait until Update() in order to guarantee that it was actually there.

wiwei commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TurnPointersOnOff.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            foreach (var icl in FindObjectsOfType<Interactable>())
+            {
+                if (icl.Dimensions == 2)
```
Nit: Add a comment on the significance of 2

julenka replied:
I instead added a property IsToggle on interactable

cre8ivepark commented:
In HoloLens 1 Mode, error messages are generated for the two-handed manipulation, when I bring up the second hand. (on cheese, coffee cup...) It might be the Bounding Box issue, not this example scene.
![2019-09-10 10_59_13-Unity 2018 4 6f1 Personal - TurnPointersOnOffExample unity - MRTK-Public-Yoon - ](https://user-images.githubusercontent.com/13754172/64638395-80375b00-d3ba-11e9-8452-dec74938eec3.png)

NullReferenceException: Object reference not set to an instance of an object
Microsoft.MixedReality.Toolkit.UI.BoundingBox.HandleProximityScaling () (at Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:2155)
Microsoft.MixedReality.Toolkit.UI.BoundingBox.Update () (at Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:1206)

julenka commented on Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs:
```diff
@@ -34,6 +34,103 @@ public class FocusProvider : BaseCoreSystem, IMixedRealityFocusProvider
+        /// <summary>
+        /// Pointer behavior for the gaze pointer.
+        /// </summary>
+        public PointerBehavior GazePointerBehavior { get; set; } = PointerBehavior.Default;
```
three reasons:
1) gazepointer actually maps to two types: ggvpointer and GazeProvider's internal gazepointer
2) the internal gazepointer actually cannot be referenced from anywhere.

I'll add a comment explaining why.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs:
```diff
@@ -209,6 +209,8 @@ public override void OnPostSceneQuery()
         {
             base.OnPostSceneQuery();
 
+            BaseCursor?.SetVisibility(IsInteractionEnabled);
```
I'll add a comment. Without this, when the poke pointer is disabled, the finger cursor still shows up. This ensures that we actually hide the finger cursor if we cannot poke things. Otherwise the system is being misleadding.

julenka commented:
Found another bug: in ggv mode example scene shows no pointer
![image](https://user-images.githubusercontent.com/168492/64639443-b675da00-d3bc-11e9-9baa-c4e7b8d8c9ca.png)

cre8ivepark commented:
Button panel layout/button size adjustments.
![2019-09-10 11_28_31-Unity 2018 4 6f1 Personal - TurnPointersOnOffExample unity - MRTK-Public-Yoon - ](https://user-images.githubusercontent.com/13754172/64640762-82e87f00-d3bf-11e9-854e-07b980d220ac.png)

julenka commented:
> In HoloLens 1 Mode, error messages are generated for the two-handed manipulation, when I bring up the second hand. (on cheese, coffee cup...) It might be the Bounding Box issue, not this example scene.
> ![2019-09-10 10_59_13-Unity 2018 4 6f1 Personal - TurnPointersOnOffExample unity - MRTK-Public-Yoon - ](https://user-images.githubusercontent.com/13754172/64638395-80375b00-d3ba-11e9-8452-dec74938eec3.png)
> 
> NullReferenceException: Object reference not set to an instance of an object
> Microsoft.MixedReality.Toolkit.UI.BoundingBox.HandleProximityScaling () (at Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:2155)
> Microsoft.MixedReality.Toolkit.UI.BoundingBox.Update () (at Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/BoundingBox/BoundingBox.cs:1206)

Seems to be fixed in my latest code, I cannot repro.

julenka commented on Assets/MixedRealityToolkit.Examples/Demos/Input/Scenes/TurnPointerOnOffExample/TurnPointersOnOff.cs:
```diff
@@ -0,0 +1,134 @@
+        private void PopulatePointerToggleList()
+        {
+            foreach (var icl in FindObjectsOfType<Interactable>())
+            {
```
Aw that was my quick prototype code :-) I'll do it proper.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs:
```diff
@@ -167,6 +169,48 @@ public virtual void UpdatePointers()
+                    {
+                        if (behavior == PointerBehavior.Off)
+                        {
+                            ptr.IsActive = false;
```
That's not quite the same behavior, since we don't want to mess with things if the behavior is default.

julenka replied:
I simplified it a bit.

julenka replied:
```
                Action<IMixedRealityPointer, PointerBehavior> setPointerState =
                    (ptr, behavior) =>
                    {
                        if (behavior == PointerBehavior.Default)
                        {
                            return;
                        }

                        bool isPointerOn = behavior == PointerBehavior.On;
                        ptr.IsActive = isPointerOn;
                        if (ptr is GenericPointer genericPtr)
                        {
                            genericPtr.IsInteractionEnabled = isPointerOn;
                        }
                        unassignedPointers.Remove(ptr);
                    };
```

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/PointerUtils.cs:
```diff
@@ -144,6 +146,131 @@ public static IEnumerable<IMixedRealityPointer> GetPointers()
+        /// <param name="handedness">Specify handedness to restrict to only right, left.</param>
+        public static void SetMotionControllerRayPointerBehavior(PointerBehavior pointerBehavior, Handedness handedness = Handedness.Any)
+        {
+            SetPointerBehavior<LinePointer>(pointerBehavior, handedness);
```
No, motion controller pointers are hand ray pointers as well :-)

julenka replied:
I added a way to differentiate the hand ray and motion controller -- you specify behavior for pointer type, input type. This will allow you to turn off motion controllers but not hand rays.

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs:
```diff
@@ -167,6 +171,49 @@ public virtual void UpdatePointers()
+
+        private void ApplyCustomPointerBehaviors()
+        {
+            if (CoreServices.InputSystem.FocusProvider is FocusProvider focusProvider)
```
It feels weird to me that the default pointer mediator, which is a thing that the focus provider uses, has to then reach all the way back out of the system and find the focus provider (and treat it directly as THE default focus provider).

I think that doing things this way avoid some breaking changes, but also still adds code cruft (i.e. generally speaking casting to specific implementations from interfaces is not really the right thing to do).

wiwei replied:
More specifically maybe it means that both:

1) The focus provider should be provided to the mediator in its constructor (or in some initialization step or in the function call itself)
2) The GetPointerBehavior shouldn't depend on the specific implementation of the focus provider - it may not be on the focus provider interface (maybe it's on should be on some IPointerBehaviorProvider interface or something)

julenka replied:
> I think that doing things this way avoid some breaking changes, but also still adds code cruft (i.e. generally speaking casting to specific implementations from interfaces is not really the right thing to do).

I thought about adding an interface, but wanted to avoid adding more interfaces. Additionally, this is a behavior specific to the input system / focus provider / mediator combo that the default MRTK uses. For example, a different system could provide this functionality in a different (simpler) way.

> It feels weird to me that the default pointer mediator, which is a thing that the focus provider uses, has to then reach all the way back out of the system and find the focus provider (and treat it directly as THE default focus provider).

I agree. I initially had the behavior set per mediator. Then I ran into the following problem: the behaviors would reset whenever I dropped and raised hands. This is because the mediator is created _per input source_. It seemed like the simplest solution was to specify the behaviors in a location that was not going to change per source, which is why I picked the focus provider.

I could try to instead have the provider specify the pointerbehavior per mediator every time the mediator is instantiated, I suppose...for some reason just grabbing the focusprovider behavior actually seems like "simpler code" to me tbh.

julenka replied:
We chatted. I will try the following approach, though if run into major issues, will probably move off of it since it can be done at later time:

```
interface IPointerPreferences
{
  // Returns the desired behavior for a given pointer
  public PointerBehavior GetPointerBehavior(IMixedRealityPointer ptr);
}
```

Then I would modify DefaultPointerMediator to take as a constructor:

```
public DefaultPointerMediator(IPointerPreferences pointerPreferences)
{
  this.pointerPreferences = pointerPreferences;
}
```

Then instead of accessing `CoreServices.InputSystem.FocusProvider` I would just look at `pointerPreferences`.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/PointerUtils.cs:
```diff
@@ -144,6 +146,131 @@ public static IEnumerable<IMixedRealityPointer> GetPointers()
+        /// <param name="handedness">Specify handedness to restrict to only right, left hands.</param>
+        public static void SetHandRayPointerBehavior(PointerBehavior pointerBehavior, Handedness handedness = Handedness.Any)
+        {
+            SetPointerBehavior<LinePointer>(pointerBehavior, handedness);
```
Good point. TeleportPointer is also a line pointer, so it's better to be more specific.

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs:
```diff
@@ -167,6 +171,49 @@ public virtual void UpdatePointers()
+                    if (pointer is GGVPointer ||
+                        pointer == CoreServices.InputSystem.GazeProvider.GazePointer)
+                    {
+                        setPointerState(pointer, focusProvider.GazePointerBehavior);
```
If we're going to have a class that gets pointer behavior, why not abstract away the complication of specific pointer interactions behind that? i.e. it feels like this should be in GetPointerBehavior, which should work for all pointers. Otherwise, would the utils also need to handle this specially as well?

wiwei replied:
Ah no, the utils do exactly the same thing - I would still say that we should have a single implementation of this, otherwise if focusProvider.GetPointerBehavior is not safe to call by itself it would open us to new bugs in the future where we forget to special case the gaze pointer.

julenka replied:
So if I'm understanding, you want a way to get the pointer behavior for the gaze pointer.

How would you make that call when the gaze pointer is an internal class in GazeProvider?

I guess we could say "if you want to get the gaze pointer behavior, you need to instead actually call GetPointerBehavior<GGVPointer>"

wiwei replied:
Ah, I think the internal gaze pointer thing is what complicates things...

I think my expectation here is that the gaze pointer would have been like other pointers, but because it's internal I guess it doesn't work out.

wiwei commented on Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs:
```diff
@@ -34,6 +34,103 @@ public class FocusProvider : BaseCoreSystem, IMixedRealityFocusProvider
+            {
+                if (kv.Key.IsAssignableFrom(pointerType))
+                {
+                    // p is subclass of key, or it is the exact class
```
What is 'p' in the context of this comment?

wiwei commented on Assets/MixedRealityToolkit.Services/InputSystem/PointerBehavior.cs:
```diff
@@ -0,0 +1,33 @@
+        /// will be always enabled. If it is not a near pointer, it will get disabled if any near pointer on the 
+        /// same hand is active. This is what allows rays to turn off when a hand is near a grabbable.
+        /// </summary>
+        Default
```
Nit: Move the Default to location zero, so that in the case that things get initialized to zero, the default is actually chosen as the default, not on.

Otherwise I think if this gets persisted and defaults get chosen things will just be "AlwaysOn"

wiwei commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
```diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
 It's possible to provide an alternate implementation of the pointer mediator by changing the [`PointerMediator`](xref:Microsoft.MixedReality.Toolkit.Input.MixedRealityPointerProfile.PointerMediator) property on the pointer profile.
 
+### How to turn pointers on and off in MRTK using PointerBehavior
+Because the pointer mediator runs every frame, it ends up controlling the active / inactive state of all pointers. Therefore, if you set a pointer's IsInteractionEnabled property in code, it will get overwritten by the pointer mediator every frame. Instead, you can specify the [`PointerBehavior`](xref:Microsoft.MixedReality.Toolkit.Inputf.PointerBehavior) to control whether pointers should be on  ...
```
```suggestion
Because the pointer mediator runs every frame, it ends up controlling the active / inactive state of all pointers. Therefore, if you set a pointer's IsInteractionEnabled property in code, it will get overwritten by the pointer mediator every frame. Instead, you can specify the [`PointerBehavior`](xref:Microsoft.MixedReality.Toolkit.Input.PointerBehavior) to control whether pointers should be on or off yourself. Note that this will only work if you are using the default [`FocusProvider`](xref:Microsoft.MixedReality.Toolkit.Input.FocusProvider) and [`DefaultPointerMediator`](xref:Microsoft.MixedReality.Toolkit.Input.DefaultPointerMediator) in MRTK. 
```

wiwei commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
````diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
+The following code will force hand rays to be on, regardless of if near a grabbable:
+
+```csharp
+// Turn off all hand rays
````
```suggestion
// Turn on all hand rays
```

wiwei commented on Documentation/Input/Pointers.md:
```diff
@@ -245,6 +245,32 @@ The [PointerResultExample scene](https://github.com/microsoft/MixedRealityToolki
 <img src="../../Documentation/Images/Input/PointerResultExample.png" style="max-width:100%;">
 
+### Turn pointers on and off
+To turn pointers on and off, set the [`PointerBehavior`]() for a given pointer type via [`PointerUtils`](). See [How to turn pointers on and off in MRTK using PointerBehavior]() for more information.
```
I think that this is missing links in PointerBehavior and PointerUtils and [How to turn pointers on and off in MRTK using PointerBehavior

julenka commented on Assets/MixedRealityToolkit/Interfaces/InputSystem/IPointerPreferences.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Provides interface for accessing pointer preferences
```
and setting

julenka commented on Assets/MixedRealityToolkit/Interfaces/InputSystem/IPointerPreferences.cs:
```diff
@@ -0,0 +1,28 @@
+    /// </summary>
+    public interface IPointerPreferences
+    {
+        PointerBehavior GetPointerBehavior(IMixedRealityPointer pointer);
```
Add doccomments

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/PointerUtils.cs:
```diff
@@ -144,6 +145,111 @@ public static IEnumerable<IMixedRealityPointer> GetPointers()
+        public static void SetHandPokePointerBehavior(PointerBehavior pointerBehavior, Handedness handedness = Handedness.Any)
+        {
+            SetPointerBehavior<PokePointer>(pointerBehavior, handedness, InputSourceType.Hand
+                );
```
Nit: Bring this to the previous line

julenka commented:
Yes, the fact that we have an internal gaze pointer that does the following:

1) has an input source as null
2) has a controller as null
3) Cannot be accessed from anything else

Is super annoying. I would love to get rid of the internal gaze pointer. Perhaps a thing for next release.

From: Will <notifications@github.com>
Sent: Wednesday, September 11, 2019 9:05 AM
To: microsoft/MixedRealityToolkit-Unity <MixedRealityToolkit-Unity@noreply.github.com>
Cc: Julia Schwarz <jschwarz@microsoft.com>; Assign <assign@noreply.github.com>
Subject: Re: [microsoft/MixedRealityToolkit-Unity] Turn rays, grab, gaze pointers on and off at runtime from code (#5920)


@wiwei commented on this pull request.

________________________________

In Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FMixedRealityToolkit-Unity%2Fpull%2F5920%23discussion_r323328721&data=02%7C01%7Cjschwarz%40microsoft.com%7C602597c5151045c1e4f708d736d1d1dc%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637038147102209028&sdata=fl%2Fu%2Bor%2Bf531Qhgh1u%2BdzA6mhj%2BAFEFqQ0jV3vyb40M%3D&reserved=0>:

> +                        {

+                            ptr.IsActive = true;

+                            if (ptr is GenericPointer genericPtr)

+                            {

+                                genericPtr.IsInteractionEnabled = true;

+                            }

+                            unassignedPointers.Remove(ptr);

+                        }

+                    };

+                foreach (IMixedRealityPointer pointer in allPointers)

+                {

+                    Handedness handedness = pointer.Controller == null ? Handedness.None : pointer.Controller.ControllerHandedness;

+                    if (pointer is GGVPointer ||

+                        pointer == CoreServices.InputSystem.GazeProvider.GazePointer)

+                    {

+                        setPointerState(pointer, focusProvider.GazePointerBehavior);

Ah, I think the internal gaze pointer thing is what complicates things...

I think my expectation here is that the gaze pointer would have been like other pointers, but because it's internal I guess it doesn't work out.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FMixedRealityToolkit-Unity%2Fpull%2F5920%3Femail_source%3Dnotifications%26email_token%3DAABJELCMYXNQLD6TJTVJU23QJEJLJA5CNFSM4IVJH67KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCEM5DKI%23discussion_r323328721&data=02%7C01%7Cjschwarz%40microsoft.com%7C602597c5151045c1e4f708d736d1d1dc%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637038147102219020&sdata=3CySWPt902B8EUTGY0tB34HhLhJF%2BT%2FFtKkLQ9WuTzA%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAABJELAQUJPGBXRNBIL75A3QJEJLJANCNFSM4IVJH67A&data=02%7C01%7Cjschwarz%40microsoft.com%7C602597c5151045c1e4f708d736d1d1dc%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637038147102229014&sdata=ebMDBVxx1FumjPunz%2BzOO1YDJpmsgJu4z%2Bz8w70FMC4%3D&reserved=0>.

julenka commented:
@wiwei and @Troy-Ferrell, please take another look. I think I've addressed all of your code & documentation comments. Some more changes I made:

- Add IPointerPreferences interface that exposes ways to get / set pointer behaviors. This allows me to have the mediator talk to an IPointerPreferences instead of grabbing the focus provider. Uses a pattern closer to dependency injection, which will hopefully help to set a pattern we can adopt to make testing easier.
- Add ability to specify pointer behavior for handedness AND input source type, something requested by @jaraymsft. This allows you to disable hand rays but keep motion controller rays active
- Added two 'watch' buttons on each wrist in the demo scene, which allows you to turn on / off rays for specific hands.

julenka commented:
Turning off the hand rays by handedness is working, but I'm not convinced it's better. Removing for now.

julenka commented:
/azp run

wiwei commented on Documentation/Input/Pointers.md:
```diff
@@ -114,6 +114,18 @@ Useful Sphere Pointer properties:
 
 <img src="../../Documentation/Images/Pointers/MRTK_Pointers_Parabolic.png" width="400">
 
+## Pointers Commonly available for Mixed Reality Platforms
```
> Pointers Commonly available for Mixed Reality Platforms [](start = 2, length = 56)

## Pointers commonly available for Mixed Reality platforms

wiwei commented on Documentation/Input/Pointers.md:
```diff
@@ -114,6 +114,18 @@ Useful Sphere Pointer properties:
 
+## Pointers Commonly available for Mixed Reality Platforms
+The following table details the pointer types that are typically used for the common platforms in MRTK. NOTE:
+it's possible to add different pointer types to these platforms. For example, you could add a Poke, Sphere pointer to VR. Additionally, VR devices with a gamepad would use the GGV pointer. 
```
>would [](start = 162, length = 5)

Should this be... "could"?

wiwei replied:
>Poke, Sphere [](start = 94, length = 12)

a PokePointer or SpherePointer to VR

wiwei commented on Documentation/Architecture/InputSystem/ControllersPointersAndFocus.md:
````diff
@@ -43,10 +43,39 @@ Because a single controller can have multiple pointers (for example, the articul
+PointerUtils.SetHandRayPointerBehavior(PointerBehavior.Default);
+```
+
+The following code will force hand rays to be on, regardless of if near a grabbable:
````
>regardless of if [](start = 50, length = 16)

"regardless of if" => regardless if

wiwei commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerBehaviorTests.cs:
```diff
@@ -0,0 +1,283 @@
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    /// <summary>
+    /// Verify that pointers can be turned on and offvia FocusProvider.SetPointerBehavior
```
>offvia [](start = 50, length = 6)

offvia -> off via

wiwei commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerBehaviorTests.cs:
```diff
@@ -0,0 +1,283 @@
+        /// <summary>
+        /// Tests that the gaze pointer can be turned on and off
+        /// </summary>
+        /// <returns></returns>
```
>/// <returns></returns> [](start = 8, length = 23)

Delete this line

wiwei replied:
(Actually, all empty returns statements - these will also fail CI, or should fail CI)

---
In reply to: [323503717](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5920#discussion_r323503717) [](ancestors = 323503717)

wiwei commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerBehaviorTests.cs:
```diff
@@ -0,0 +1,283 @@
+            yield return rightHand.Show(Vector3.zero);
+            yield return leftHand.Show(Vector3.zero);
+
+            TestContext.Out.WriteLine("Show both hands");
```
>TestContext.Out.WriteLine("Show both hands"); [](start = 12, length = 45)

Nit: I think this WriteLine should be a few lines above (when it is showing the hands).

wiwei commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerBehaviorTests.cs:
```diff
@@ -0,0 +1,283 @@
+    {
+        private class PointerStateContainer
+        {
+            public bool? LinePointerEnabled { get; set; }
```
Nit: Add a comment about the tribit-nature of this - this doesn't actually map to Default, AlwaysOn, AlwaysOff, but actually is more about, it's non-null validating that the current active state matches the bool value.

wiwei commented on Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs:
```diff
@@ -1407,5 +1412,126 @@ public void OnSpeechKeywordRecognized(SpeechEventData eventData)
+            return GetPointerBehavior(typeof(T), handedness, sourceType);
+        }
+
+        private PointerBehavior GetPointerBehavior(Type type, Handedness handedness, InputSourceType sourceType)
```
>GetPointerBehavior [](start = 32, length = 18)

If we expect that GetPointerBehaviors can be called in the Update() loop, consider avoiding LINQ:

[https://docs.microsoft.com/en-us/windows/mixed-reality/performance-recommendations-for-unity#avoid-expensive-operations](https://docs.microsoft.com/en-us/windows/mixed-reality/performance-recommendations-for-unity#avoid-expensive-operations)

wiwei replied:
Same for SetPointerBehavior

---
In reply to: [323507446](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5920#discussion_r323507446) [](ancestors = 323507446)

wiwei commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/PointerUtils.cs:
```diff
@@ -144,6 +146,128 @@ public static IEnumerable<IMixedRealityPointer> GetPointers()
+        private static void WarnAboutSettingCustomPointerBehaviors()
+        {
+            Debug.LogWarning("Setting custom pointer behaviors only works if the input system is using the default MRTK focus provider. " +
+                "Are you using a custom Focus Provider?");
```
>Focus Provider [](start = 40, length = 14)

Nit: "Are you using a custom Focus Provider that doesn't implement IPointerPreferences"

To give folks who see the error a next step on what to do (i.e. implement IPointerPreferences)

wiwei commented on Assets/MixedRealityToolkit/Interfaces/InputSystem/IPointerPreferences.cs:
```diff
@@ -0,0 +1,50 @@
+        /// handedness, and input type
+        /// </summary>
+        /// <typeparam name="T">All pointers that are of this type, or a subclass of this type, will have the given behavior</typeparam>
+        /// <param name="handedness">Specify Handedness.Any to apply to all handedness, or specify a specific handedness to just disable, right, left.</param>
```
>Any [](start = 56, length = 3)

Just out of curiousity, does Both also work? I know that this enum has been confusing in the past with some other APIs.

julenka replied:
Yes it does.

wiwei commented on Assets/MixedRealityToolkit/Interfaces/InputSystem/IPointerPreferences.cs:
```diff
@@ -0,0 +1,50 @@
+        /// </summary>
+        /// <typeparam name="T">All pointers that are of this type, or a subclass of this type, will have the given behavior</typeparam>
+        /// <param name="handedness">Specify Handedness.Any to apply to all handedness, or specify a specific handedness to just disable, right, left.</param>
+        /// <param name="sourceType">Allows specification of pointer behavior per input source, so that pointers can be disabled for hands but not controllers</param>
```
>controllers [](start = 147, length = 11)

Nit: ... and vice versa.

wiwei replied:
Just to make sure it's clear that it's to allow you to customize both (or between other input sources, not just this relationship between controllers). (Or maybe adding "for example...")

wiwei commented on Assets/MixedRealityToolkit/Definitions/InputSystem/PointerBehavior.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    /// <summary>
+    /// Specifies how a pointer in MRTK's default input system behaves.
+    /// By default, pointer behaviors are set to Default.
```
>/// By default, pointer behaviors are set to Default. [](start = 4, length = 53)

Super nit: I think this line doesn't add that much (I think it's expected that something that is named "default" is the default)

wiwei replied:
I think maybe if the enum was called "NotTheDefaultJustKiddingItsActuallyTheDefault" then having this comment would be good (actually in general I think that it's a good callout, just not in this case because the naming is fairly obvious)

---
In reply to: [323509091](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5920#discussion_r323509091) [](ancestors = 323509091)

pampas93 commented:
Is it the expected behavior to not receive any `IMixedRealityPointerHandler` callbacks after switching off HandRayPointer? using the `PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff);` ?

sostel commented:
> Is it the expected behavior to not receive any `IMixedRealityPointerHandler` callbacks after switching off HandRayPointer?
> using `PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff);`

If you turn off the HandRayPointer and you may have to enable an alternative pointer. For example:

public void DisableHandRay()
        {
            PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff, Handedness.Any);
            PointerUtils.SetGazePointerBehavior(PointerBehavior.Default);
        }

pampas93 commented:
> 
> 
> > Is it the expected behavior to not receive any `IMixedRealityPointerHandler` callbacks after switching off HandRayPointer?
> > using `PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff);`
> 
> If you turn off the HandRayPointer and you may have to enable an alternative pointer. For example:
> 
> ```
> public void DisableHandRay()
>         {
>             PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff, Handedness.Any);
>             PointerUtils.SetGazePointerBehavior(PointerBehavior.Default);
>         }
> ```

With the GazePointer enabled and the HandRayPointer disabled, I still get the same behavior. No callbacks to the IMixedRealityPointerHandler functions.

For example, the log is never printed (with the below type of code). With the class implementing IMixedRealityPointerHandler interface. But commenting out the SetHandRayPointerBehavior works fine.
void OnEnable()
{
    PointerUtils.SetHandRayPointerBehavior(PointerBehavior.AlwaysOff, Handedness.Any);
    PointerUtils.SetGazePointerBehavior(PointerBehavior.Default);
}

public void OnPointerDown(MixedRealityPointerEventData eventData)
{
    Debug.Log("Pointer down");
}

Troy-Ferrell commented:
@pampas93 how are you activating the pointer down event with the gaze pointer?

pampas93 commented:
> 
> 
> @pampas93 how are you activating the pointer down event with the gaze pointer?

I'm implementing the `IMixedRealityPointerHandler` and registering the handler with the `CoreServices.InputSystem` within the OnEnable of the script

Troy-Ferrell commented:
@pampas93 no sorry I meant at runtime,are you just running in editor with gaze provided or on device? Do you look at the item with the script and say select with voice?

pampas93 commented:
I'm running this on the editor, with the hand visualized (input simulation). But I didn't get your second question. I'm not using any interactable script on the item to have it say select with voice.

Troy-Ferrell commented:
@pampas93 Ok I think I know what is going on. The reason why you are not getting a OnPointerDown event is because there is no pointer to fire it. 

You disable the hand ray pointer via PointerUtils.SetHandRayPointerBehavior() so that will no longer fire the events or be an active pointer. 

The GazePointer (GGVPointer) actually isn't active by default in editor. This is because by default the pointer is only created for "GGV Hand" input sources/controllers. 
![image](https://user-images.githubusercontent.com/25975362/73386703-27dafa00-4284-11ea-8bd6-38fa33e82789.png)

I know this is a bit weird, but basically this type of "GGV Hand" will only show up on HL1. In editor and HL2, the type of hand created is "Articulated Hand" which only spawns the hand ray pointer and grab/poke pointers. 

I created a new issue #7182 to track making this more apparent in the API

dahaideyu commented:
How to close TeleportPointer  and ParabolicTeleportPointer  ，Use code is not work
			PointerUtils.SetPointerBehavior<TeleportPointer>(PointerBehavior.AlwaysOff);

			PointerUtils.SetPointerBehavior<ParabolicTeleportPointer>(PointerBehavior.AlwaysOff); @julenka
